gradoop-flink/src/main/java/org/gradoop/flink/algorithms/fsm/dimspan/DIMSpan.java

```diff
@@ -0,0 +1,373 @@
+      comparator =  new ProportionalLabelComparator();
+    } else if (fsmConfig.getDictionaryType() == DictionaryType.INVERSE_PROPORTIONAL) {
+      comparator =  new InverseProportionalLabelComparator();
+    } else {
+      comparator = new AlphabeticalLabelComparator();
+    }
+  }
+
+  /**
+   * Executes the DIMSpan algorithm.
+   * Orchestration of preprocessing, mining and postprocessing.
+   *
+   * @param input input graph collection
+   * @return frequent patterns
+   */
+  public DataSet<GraphTransaction> execute(DataSet<LabeledGraphStringString> input) {
+
+    DataSet<int[]> encodedInput = preProcess(input);
+    DataSet<int[]> encodedOutput = mine(encodedInput);
+
+    return postProcess(encodedOutput);
+  }
+
+  /**
+   * Triggers the label-frequency base preprocessing
+   *
+   * @param graphs input
+   * @return preprocessed input
+   */
+  private DataSet<int[]> preProcess(DataSet<LabeledGraphStringString> graphs) {
+
+    // Determine cardinality of input graph collection
+    this.graphCount = Count
+      .count(graphs);
+
+    // Calculate minimum frequency
+    this.minFrequency = graphCount
+      .map(new MinFrequency(fsmConfig));
+
+    // Execute vertex label pruning and dictionary coding
+    DataSet<LabeledGraphIntString> graphsWithEncodedVertices = encodeVertices(graphs);
+
+    // Execute edge label pruning and dictionary coding
+    DataSet<int[]> encodedGraphs = encodeEdges(graphsWithEncodedVertices);
+
+    // return all non-obsolete encoded graphs
+    return encodedGraphs
+      .filter(new NotEmpty());
+  }
+
+  /**
+   * Triggers the iterative mining process.
+   *
+   * @param graphs preprocessed input graph collection
+   * @return frequent patterns
+   */
+  protected DataSet<int[]> mine(DataSet<int[]> graphs) {
+
+    DataSet<GraphWithPatternEmbeddingsMap> searchSpace = graphs
+      .map(new InitSingleEdgePatternEmbeddingsMap(gSpan, fsmConfig));
+
+    // Workaround to support multiple data sinks: create pseudo-graph (collector),
+    // which embedding map will be used to union all k-edge frequent patterns
+    DataSet<GraphWithPatternEmbeddingsMap> collector = graphs
+      .getExecutionEnvironment()
+      .fromElements(true)
+      .map(new CreateCollector());
+
+    searchSpace = searchSpace.union(collector);
+
+    // ITERATION HEAD
+
+    IterativeDataSet<GraphWithPatternEmbeddingsMap> iterative = searchSpace
+      .iterate(MAX_ITERATIONS);
+
+    // ITERATION BODY
+
+    DataSet<WithCount<int[]>> reports = iterative
+      .flatMap(new ReportSupportedPatterns());
+
+    DataSet<int[]> frequentPatterns = getFrequentPatterns(reports);
+
+    DataSet<GraphWithPatternEmbeddingsMap> grownEmbeddings = iterative
+      .map(new GrowFrequentPatterns(gSpan, fsmConfig))
+      .withBroadcastSet(frequentPatterns, DIMSpanConstants.FREQUENT_PATTERNS)
+      .filter(new NotObsolete());
+
+    // ITERATION FOOTER
+
+    return iterative
+      .closeWith(grownEmbeddings, frequentPatterns)
```

Comment:
1. Is this a variable broadcast?
2. Does this “closure” applies to the `IsFrequentPatternCollector` and `ExpandFrequentPatterns` object?
3. If conditions 1 and 2 hold, why is this method used instead of BroadcastWith?
4. If conditions 1, 2 and 3 hold, why didn't you simply pass the elements within the classes' constructor? Is this more efficient?